### PR TITLE
Added Openstack security group caching.

### PR DESCRIPTION
**Pull request by varikin**
Closes spinnaker/spinnaker#969

### CODE REVIEW

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/securitygroups/Rule.groovy [hunk 1]
```diff
@@ -25,6 +26,7 @@ import groovy.transform.Sortable
  * @see IpRangeRule
  * @see SecurityGroupRule
  */
+@JsonTypeInfo(use=JsonTypeInfo.Id.CLASS, include=JsonTypeInfo.As.PROPERTY, property="class")
```
Comment (derekolk) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/securitygroups/Rule.groovy]: Use single string instead of gstring.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/Keys.groovy [hunk 2]
```diff
@@ -87,6 +88,18 @@ class Keys {
               result << [id: parts[2], account: parts[3], region: parts[4]]
             }
             break
+          case Namespace.SECURITY_GROUPS.ns:
+            if (parts.length == 6) {
+              def names = Names.parseName(parts[2])
+              result << [
+                application: names.app,
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/Keys.groovy]: All of the other formatting is done on a single line ... Why the change?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackSecurityGroup.groovy [hunk 3]
```diff
@@ -37,8 +46,45 @@ class OpenstackSecurityGroup implements SecurityGroup {
   final Set<Rule> inboundRules
   final Set<Rule> outboundRules
 
+  @JsonIgnore
   @Override
   SecurityGroupSummary getSummary() {
     new OpenstackSecurityGroupSummary(name: name, id: id)
   }
+
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackSecurityGroup.groovy]: I have been changing this to use a builder ... Just FYI.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgent.groovy [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  OpenstackSecurityGroupCachingAgent(OpenstackNamedAccountCredentials account, String region, ObjectMapper objectMapper) {
+    super(account, region)
+    this.objectMapper = objectMapper
+  }
+
+  @Override
+  CacheResult loadData(ProviderCache providerCache) {
+    CacheResultBuilder cacheResultBuilder = new CacheResultBuilder()
+    clientProvider.getSecurityGroups(region).each { securityGroup ->
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgent.groovy]: Could we run into an NPE if client provider returns no security groups for the region?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy [hunk 5]
```diff
@@ -17,55 +17,34 @@
...
-            provider.getSecurityGroups(region)
-              .flatten()
-              .collect { convertToOpenstackSecurityGroup(includeRules, it, account.name, region) }
-          )
-        }
-      }
-    }
-
-    securityGroups
+  OpenstackSecurityGroupProvider(Cache cacheView, ObjectMapper objectMapper) {
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:45]: Do we want to make the parameters final too?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy [hunk 6]
```diff
@@ -75,67 +54,55 @@ class OpenstackSecurityGroupProvider implements SecurityGroupProvider<OpenstackS
 
   @Override
   Set<OpenstackSecurityGroup> getAll(boolean includeRules) {
-    getSecurityGroups(includeRules)
+    getAllMatchingKeyPattern(Keys.getSecurityGroupKey('*', '*', '*', '*'), includeRules)
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:57]: Should we make a wildcard a constant and DRY this up?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy [hunk 7]
```diff
@@ -75,67 +54,55 @@ class OpenstackSecurityGroupProvider implements SecurityGroupProvider<OpenstackS
...
-  Set<OpenstackSecurityGroup> getAllByAccountAndRegion(boolean includeRule, String account, String region) {
-    getAllByAccount(includeRule, account).findAll { it.region == region }
+  Set<OpenstackSecurityGroup> getAllByAccountAndRegion(boolean includeRules, String account, String region) {
+    getAllMatchingKeyPattern(Keys.getSecurityGroupKey('*', '*', account, region), includeRules)
   }
 
   @Override
   OpenstackSecurityGroup get(String account, String region, String name, String vpcId) {
-    getAllByAccountAndRegion(true, account, region).find { it.name == name }
+    getAllMatchingKeyPattern(Keys.getSecurityGroupKey(name, '*', account, region), true)[0]
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:82]: Groovy has this cool operator list?.first() rather than [0] ... Looks cleaner :)

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgentSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    provider = Mock(OpenstackClientProvider)
+    credentials = GroovyMock(OpenstackCredentials)
+    credentials.provider >> provider
+    namedAccountCredentials = Mock(OpenstackNamedAccountCredentials)
+    namedAccountCredentials.credentials >> credentials
+    objectMapper = Mock(ObjectMapper)
+    cachingAgent = Spy(OpenstackSecurityGroupCachingAgent, constructorArgs: [namedAccountCredentials, region, objectMapper])
+    _ * cachingAgent.getAccountName() >> accountName
+  }
+
```
Comment (derekolk) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgentSpec.groovy:63]: What about exception test case?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 5 above)
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:45]: What does that buy us?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 6 above)
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:57]: I was copying the GoogleSecurityGroupProvider.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 7 above)
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:82]: `first()` throws an exception on an empty list.

Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:82]: Yeah ... that would change the behavior.  Disregard my comment.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 6 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:57]: Cool ... Then I would leave as is.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 5 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:45]: Just ensures that constructor args can't be changed, so in our case nothing.  It would be more of a conventional thing and since we aren't doing that all over the place, lets forget it.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgent.groovy (hunk 4 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgent.groovy]: I believe each region always has at least 'Default'.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 7 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:82]: What happens if getAllMatchingKeyPattern returns null? Is that possible?

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProviderSpec.groovy [hunk 9]
```diff
@@ -52,10 +54,8 @@ class OpenstackSecurityGroupProviderSpec extends Specification {
   Set<OpenstackSecurityGroup> allSecurityGroups = account1East.plus(account1West).plus(account2East).plus(account2West)
 
   def setup() {
-    accountCredentialsProvider = Mock()
-
-    // TODO Shouldn't need the spy once the data is coming from the caching layer
-    provider = Spy(OpenstackSecurityGroupProvider, constructorArgs: [accountCredentialsProvider])
+    provider = new OpenstackSecurityGroupProvider(cache, mapper)
+    cache.mergeAll(Keys.Namespace.SECURITY_GROUPS.ns, getAllCacheData())
   }
```
Comment (drmaas) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProviderSpec.groovy:61]: Add exception test cases

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgent.groovy (hunk 4 above)
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgent.groovy]: I believe the Openstack4J call returns an empty list, but I can't be positive. I added a `?.` to safe guard this.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy (hunk 7 above)
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy:82]: Just returns an empty collection, but I will add a test for that case to be sure.

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProviderSpec.groovy (hunk 9 above)
Comment (varikin) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProviderSpec.groovy:61]: What sort of exception test cases do you want? I am neither throwing or handling any checked exceptions in this class.

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgentSpec.groovy (hunk 8 above)
Comment (varikin) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgentSpec.groovy:63]: What do you want tested for exceptions? `OpenstackProviderException` could be thrown, but it will just bubble up.

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProviderSpec.groovy (hunk 9 above)
Comment (drmaas) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProviderSpec.groovy:61]: from the network spec - 

```
  void "test get all - exception"() {
    given:
    Collection<String> filters = Mock(Collection)
    Throwable throwable = new JedisException('test')

    when:
    provider.getAll()

    then:
    1 * cache.filterIdentifiers(NETWORKS.ns, "${OpenstackCloudProvider.ID}:${NETWORKS.ns}:*:*:*") >> filters
    1 * cache.getAll(NETWORKS.ns, filters, _) >> { throw throwable }
    JedisException exception = thrown(JedisException)
    exception == throwable
  }
```

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgentSpec.groovy (hunk 8 above)
Comment (varikin) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgentSpec.groovy:63]: Added test

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProviderSpec.groovy (hunk 9 above)
Comment (varikin) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProviderSpec.groovy:61]: Added
